addons/project/models/project.py

```diff
@@ -893,15 +900,23 @@ def _send_rating_all(self):
     # Privacy
     # ---------------------------------------------------
 
-    def _change_privacy_visibility(self):
+    def _change_privacy_visibility(self, new_visibility):
         """
         Unsubscribe non-internal users from the project and tasks if the project privacy visibility
-        goes to a value different than 'portal'
+        goes from 'portal' to a different value.
+        If the privacy visibility is set to 'portal', subscribe back project and tasks partners.
         """
-        for project in self.filtered(lambda p: p.privacy_visibility != 'portal'):
-            portal_users = project.message_partner_ids.user_ids.filtered('share')
-            project.message_unsubscribe(partner_ids=portal_users.partner_id.ids)
-            project.mapped('tasks')._change_project_privacy_visibility()
+        for project in self:
+            if project.privacy_visibility == new_visibility:
+                continue
+            if new_visibility == 'portal':
+                project.message_subscribe(partner_ids=project.partner_id.ids)
+                for task in project.mapped('task_ids').filtered('partner_id'):
+                    task.message_subscribe(partner_ids=task.partner_id.ids)
+            elif project.privacy_visibility == 'portal':
+                portal_users = project.message_partner_ids.user_ids.filtered('share')
+                project.message_unsubscribe(partner_ids=portal_users.partner_id.ids)
+                project.mapped('task_ids')._unsubscribe_portal_users()
```

Comment:
Why do we just remove the portal user on the opened tasks and not all the tasks in the project? 🤔

Reply:
Fixed